File: crack-data-structures-and-algorithms/leetcode/python-impl/find_first_and_last_position_of_element_in_sorted_array_q34.py
# -*- coding: utf-8 -*-
# 0xCCCCCCCC

def search_range(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: List[int]
    """
    lb = lower_bound(nums, target)
    if lb == len(nums) or nums[lb] != target:
        # No element matched, simply prune.
        return [-1, -1]

    rb = upper_bound(nums, target)
    if rb - 1 >= 0 and nums[rb-1] != target:
        return [-1, -1]
    else:
        rb -= 1

    return [lb, rb]

def lower_bound(nums, target):
    count = len(nums)
    first = 0
    while count > 0:
        half = count // 2
        mid = first + half
        if nums[mid] < target:
            first = mid + 1
            count -= half + 1
        else:
            count = half
    return first

def upper_bound(nums, target):
    count = len(nums)
    first = 0
    while count > 0:
        half = count // 2
        mid = first + half
        if not (target < nums[mid]):
            first = mid + 1
            count -= half + 1
        else:
            count = half
    return first

nums = [5,7,7,8,8,10]
target = 8
print(search_range(nums, target) == [3, 4])

nums = [5,7,7,8,8,10]
target = 6
print(search_range(nums, target) == [-1, -1])

nums = [2,4,6,8,10]
target = 6
print(search_range(nums, target) == [2, 2])

nums = [1]
target = 1
print(search_range(nums, target) == [0, 0])

nums = []
target = 1
print(search_range(nums, target) == [-1, -1])

File: crack-data-structures-and-algorithms/leetcode/find_first_and_last_position_of_element_in_sorted_array_q34.cpp
#include <functional>
#include <vector>

using namespace std;

// 核心思路
// 直接使用 lower_bound & upper_bound 解决
// 注意这里复用一套框架，抽象了 predicate
// 另注意这里的 predicate 和 STL 里用的 predicate 语义作用都不一样
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        auto lb = search_bound(nums, target, [](int val, int target) { return val >= target; });
        auto ub = search_bound(nums, target, [](int val, int target) { return val > target; });
        return lb == nums.size() || nums[lb] != target ? vector<int>{-1, -1} : vector<int>{lb, ub - 1};
    }

    int search_bound(const vector<int>& nums, int target, function<bool(int val, int target)> pred) {
        int count = nums.size();
        int first = 0;
        while (count > 0) {
            int half = count / 2;
            int mid = first + half;
            if (!pred(nums[mid], target)) {
                first = mid + 1;
                count -= half + 1;
            } else {
                count = half;
            }
        }

        return first;
    }
};
